stability: panic in ts.unionIterator.max

`beta` cluster running sha f2a807a

Cluster was painfully slow, and while we were poking around the admin page, one of the nodes (`104.196.141.29`) paniced in the timeseries code.

Looping in @MattTracy, but I'll take a look since he's away for a bit.

```
I160803 14:00:00.285380 storage/replica_trigger.go:308  store=2:2 range=5375 [/System/tsd/cr.node.exec.error-10m/4/10s/2016-07-03T16:00:00Z-/System/tsd/cr.node.exec.error-1h/6/10s/2016-07-03T10:00:00Z): new range lease replica {2 2 15} 2016-08-03 13:59:58.111658408 +0000 UTC 6.434032905s following replica {5 5 13} 2016-08-03 13:58:29.424043514 +0000 UTC 1m28.687614894s [physicalTime=2016-08-03 14:00:00.285277479 +0000 UTC]
W160803 14:00:04.479294 raft/raft.go:593  [range 4449] 23 stepped down to follower since quorum is not active
W160803 14:00:04.480936 raft/raft.go:593  [range 3330] 13 stepped down to follower since quorum is not active
W160803 14:00:04.481096 raft/raft.go:593  [range 718] 2f stepped down to follower since quorum is not active
W160803 14:00:04.484603 storage/replica.go:1790  failed to look up recipient replica 37 in range 580 while sending MsgVoteResp: replica 37 not present in {3 3 36}, []
panic: runtime error: index out of range

goroutine 1801632672 [running]:
panic(0x1969ba0, 0xc82000e040)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/ts.unionIterator.max(0x2b06768, 0x0, 0x0, 0xc8204c4050)
        /go/src/github.com/cockroachdb/cockroach/ts/query.go:387 +0x101
github.com/cockroachdb/cockroach/ts.(unionIterator).(github.com/cockroachdb/cockroach/ts.max)-fm(0x2b06768)
        /go/src/github.com/cockroachdb/cockroach/ts/query.go:511 +0x38
github.com/cockroachdb/cockroach/ts.(*DB).Query(0xc8204c8018, 0xc83a526d00, 0x17, 0xc8748b1d20, 0xc8748b1d24, 0xc8748b1d28, 0x0, 0x0, 0x0, 0x1, ...)
        /go/src/github.com/cockroachdb/cockroach/ts/query.go:524 +0x9a2
github.com/cockroachdb/cockroach/ts.(*Server).Query(0xc82009a808, 0x7ff3aec979e8, 0xc88a08c9c0, 0xc88a08ca20, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/ts/server.go:72 +0x2b6
github.com/cockroachdb/cockroach/ts/tspb._TimeSeries_Query_Handler(0x19d2a00, 0xc82009a808, 0x7ff3aec979e8, 0xc88a08c9c0, 0xc8af1fcbe0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/ts/tspb/timeseries.pb.go:349 +0x168
google.golang.org/grpc.(*Server).processUnaryRPC(0xc8204cc000, 0x7ff3aebb9b90, 0xc825a8d4d0, 0xc892e43b30, 0xc82017b710, 0x2785fb0, 0xc88a08c990, 0x0, 0x0)
        /go/src/google.golang.org/grpc/server.go:608 +0x10d2
google.golang.org/grpc.(*Server).handleStream(0xc8204cc000, 0x7ff3aebb9b90, 0xc825a8d4d0, 0xc892e43b30, 0xc88a08c990)
        /go/src/google.golang.org/grpc/server.go:766 +0x10a0
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc8217cd950, 0xc8204cc000, 0x7ff3aebb9b90, 0xc825a8d4d0, 0xc892e43b30)
        /go/src/google.golang.org/grpc/server.go:419 +0xa0
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/src/google.golang.org/grpc/server.go:420 +0x9a
```

Full logs since start:
[latest_logs.tar.gz](https://github.com/cockroachdb/cockroach/files/399610/latest_logs.tar.gz)

--- a/ts/query.go
+++ b/ts/query.go
@@ -515,6 +515,11 @@ func (db *DB) Query(query tspb.Query, r Resolution, startNanos, endNanos int64) 
 	// unionIterator at each offset encountered. If the query is requesting a
 	// derivative, a rate of change is recorded instead of the actual values.
 	iters.init()
+	if !iters.isValid() {
+		// We have no data to return.
+		return nil, sources, nil
+	}
+
 	var last tspb.TimeSeriesDatapoint
 	if isDerivative {
 		last = tspb.TimeSeriesDatapoint{

--- a/ts/query_test.go
+++ b/ts/query_test.go
@@ -317,6 +317,13 @@ func (tm *testModel) assertQuery(
 		}
 	}
 
+	// Verify that expected sources match actual sources.
+	sort.Strings(expectedSources)
+	sort.Strings(actualSources)
+	if !reflect.DeepEqual(actualSources, expectedSources) {
+		tm.t.Error(errors.Errorf("actual source list: %v, expected: %v", actualSources, expectedSources))
+	}
+
 	// Iterate over data in all dataSpans and construct expected datapoints.
 	var startOffset int32
 	isDerivative := q.GetDerivative() != tspb.TimeSeriesQueryDerivative_NONE
@@ -333,6 +340,12 @@ func (tm *testModel) assertQuery(
 	}
 
 	iters.init()
+	if !iters.isValid() {
+		if a, e := 0, len(expectedDatapoints); a != e {
+			tm.t.Error(errors.Errorf("query had zero datapoints, expected: %v", expectedDatapoints))
+		}
+		return
+	}
 	currentVal := func() tspb.TimeSeriesDatapoint {
 		var value float64
 		switch q.GetSourceAggregator() {
@@ -380,11 +393,6 @@ func (tm *testModel) assertQuery(
 		iters.advance()
 	}
 
-	sort.Strings(expectedSources)
-	sort.Strings(actualSources)
-	if !reflect.DeepEqual(actualSources, expectedSources) {
-		tm.t.Error(errors.Errorf("actual source list: %v, expected: %v", actualSources, expectedSources))
-	}
 	if !reflect.DeepEqual(actualDatapoints, expectedDatapoints) {
 		tm.t.Error(errors.Errorf("actual datapoints: %v, expected: %v", actualDatapoints, expectedDatapoints))
 	}
@@ -457,8 +465,37 @@ func TestQuery(t *testing.T) {
 	// Test with everything specified.
 	tm.assertQuery("test.multimetric", nil, tspb.TimeSeriesQueryAggregator_MIN.Enum(), tspb.TimeSeriesQueryAggregator_MAX.Enum(),
 		tspb.TimeSeriesQueryDerivative_NON_NEGATIVE_DERIVATIVE.Enum(), resolution1ns, 0, 90, 8, 2)
-	// Test query that returns no data.
-	tm.assertQuery("nodata", nil, nil, nil, nil, resolution1ns, 0, 90, 0, 0)
+
+	// Test queries that return no data. Check with every
+	// aggregator/downsampler/derivative combination. This situation is
+	// particularly prone to nil panics (parts of the query system will not have
+	// data).
+	aggs := []tspb.TimeSeriesQueryAggregator{
+		tspb.TimeSeriesQueryAggregator_MIN, tspb.TimeSeriesQueryAggregator_MAX,
+		tspb.TimeSeriesQueryAggregator_AVG, tspb.TimeSeriesQueryAggregator_SUM,
+	}
+	derivs := []tspb.TimeSeriesQueryDerivative{
+		tspb.TimeSeriesQueryDerivative_NONE, tspb.TimeSeriesQueryDerivative_DERIVATIVE,
+		tspb.TimeSeriesQueryDerivative_NON_NEGATIVE_DERIVATIVE,
+	}
+	for _, downsampler := range aggs {
+		for _, agg := range aggs {
+			for _, deriv := range derivs {
+				tm.assertQuery(
+					"nodata",
+					nil,
+					downsampler.Enum(),
+					agg.Enum(),
+					deriv.Enum(),
+					resolution1ns,
+					0,
+					90,
+					0,
+					0,
+				)
+			}
+		}
+	}
 
 	// Verify querying specific sources, thus excluding other available sources
 	// in the same time period.

